### PR TITLE
fixed last_upload and last_download in storage

### PR DESCRIPTION
**Pull request by isaingard**
created test_no_download, fixed last_upload and last_download

### CODE REVIEW

File: storage.py [hunk 1]
```diff
@@ -176,8 +177,16 @@ def delete_all_local_and_remote(self, create_dir=True):
...
 class iPad(Device):
 	def synchronize(self):
-		self.sync_delete_files()
-		self.sync_download_files()
+		if self.local_file_exists(last_download_filename):
+			local_last_time = self.get_local_modified_time(last_download_filename)
+		else:
+			local_last_time = datetime.datetime.fromtimestamp(0, datetime.timezone.utc).replace(microsecond=0)
+		remote_last_time = self.get_remote_modified_time(last_upload_filename)
+		if remote_last_time > local_last_time:
```
Comment (michael-hewitt) [inline storage.py]: Please bury this low level code in a single method that returns True or False.

File: storage.py [hunk 2]
```diff
@@ -222,17 +246,28 @@ def test_delete():
...
 	@staticmethod
 	def test():
+		iPad.test_no_download()
 		iPad.test_download()
 		iPad.test_delete()
 
 class Laptop(Device):
 	def synchronize(self):
+		if not self.local_file_exists(last_upload_filename):
+			os.open(self.local_dir + '/' + last_upload_filename, os.O_CREAT)
```
Comment (michael-hewitt) [inline storage.py]: No need to create local last_upload file

File: storage.py [hunk 3]
```diff
@@ -222,17 +246,28 @@ def test_delete():
...
+		self.upload()
```
Comment (michael-hewitt) [inline storage.py]: make sure this function does not upload last_upload

File: storage.py [hunk 4]
```diff
@@ -222,17 +246,28 @@ def test_delete():
...
+		self.delete()
+		time.sleep(1)
+		os.utime(self.local_dir + '/' + last_upload_filename, (datetime.datetime.timestamp(datetime.datetime.now()),
+									   datetime.datetime.timestamp(datetime.datetime.now())))
```
Comment (michael-hewitt) [inline storage.py]: No reason to mess with a local last_upload.  Instead write to remote last_upload here.

File: storage.py (hunk 3 above)
Comment (michael-hewitt) [inline storage.py]: Make sure there is no last_upload file in our songs or tunes repository.

File: storage.py [hunk 5]
```diff
@@ -286,12 +320,11 @@ def test_newer_local_song():
 		assert remote_time < local_time
 		laptop.synchronize()
 		for local_file in os.listdir(test_dir):
-			local_content = laptop.get_local_file_content(local_file)
-			for remote_file in service.list_directories_and_files(share, test_dir):
-				if remote_file.name != last_upload_filename:
-					assert local_file == remote_file.name
-					remote_content = laptop.get_remote_file_content(remote_file)
-					assert remote_content == local_content
+			if not local_file == last_upload_filename:
```
Comment (michael-hewitt) [inline storage.py]: Convert this to an assertion.   There should *never* be a local last_upload file on the laptop.

File: storage.py [hunk 6]
```diff
@@ -304,14 +337,13 @@ def test_older_local_song():
...
-		Laptop(test_dir, test_dir).synchronize()
+		laptop.synchronize()
 		for local_file in os.listdir(test_dir):
-			local_content = laptop.get_local_file_content(local_file)
-			for remote_file in service.list_directories_and_files(share, test_dir):
-				if remote_file.name != last_upload_filename:
-					assert local_file == remote_file.name
-					remote_content = laptop.get_remote_file_content(remote_file)
-					assert remote_content != local_content
+			if not local_file == last_upload_filename:
```
Comment (michael-hewitt) [inline storage.py]: Convert this 'if' to an assert.  There should never be a local last_upload file.

**Comment by isaingard** [comment]
more changes added

File: storage.py [hunk 7]
```diff
@@ -259,13 +262,11 @@ def test():
...
 	def synchronize(self):
-		if not self.local_file_exists(last_upload_filename):
-			os.open(self.local_dir + '/' + last_upload_filename, os.O_CREAT)
 		self.upload()
 		self.delete()
-		time.sleep(1)
-		os.utime(self.local_dir + '/' + last_upload_filename, (datetime.datetime.timestamp(datetime.datetime.now()),
-									   datetime.datetime.timestamp(datetime.datetime.now())))
+		if self.remote_file_exists(last_upload_filename):
+			self.delete_remote_file(last_upload_filename)
```
Comment (michael-hewitt) [inline storage.py]: We shouldn't need to delete this file because the self.delete will take care of it.  Let's just assert that it isn't there instead.

**Comment by michael-hewitt** [comment]
Please replace the code to delete the remote last_upload file with an assert that it doesn't exist.  Then squash all your commits into one and force push up to GitHub.

**Pull request merged**